File: GMUtilsAndroid/GMUtils/GMUtils/src/main/java/gmutils/firebase/storage/FirebaseStorage.java
package gmutils.firebase.storage;

import android.net.Uri;

import com.google.firebase.storage.StorageReference;
import com.google.firebase.storage.UploadTask;

import org.jetbrains.annotations.Nullable;

import java.io.File;
import java.io.IOException;

import gmutils.DateOp;
import gmutils.StringSet;
import gmutils.firebase.Response;
import gmutils.listeners.ResultCallback;

public class FirebaseStorage implements IFirebaseStorage {
    public final com.google.firebase.storage.FirebaseStorage storage;

    public FirebaseStorage() {
        try {
            Class.forName("com.google.firebase.storage.FirebaseStorage");
        } catch (ClassNotFoundException e) {
            e.printStackTrace();
            throw new IllegalStateException("add this line to gradle script file:\n" +
                    "//https://firebase.google.com/docs/database/android/start\n" +
                    "implementation 'com.google.firebase:firebase-storage:19.6.0'");
        }
        storage = com.google.firebase.storage.FirebaseStorage.getInstance();
    }

    /*private String _refinePath(String path) {
        try {
            path = FirebaseUtils.refinePathFragmentNames(path);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
        return path;
    }*/

    //----------------------------------------------------------------------------

    @Override
    public void upload(Uri fileUri, String toPath, ResultCallback<Response<String>> callback) {
        //toPath = _refine Path(toPath);
        var ref = storage.getReference(toPath);
        upload(fileUri, ref, callback);
    }

    public void upload(
            Uri fileUri,
            StorageReference reference,
            ResultCallback<Response<String>> callback
    ) {
        UploadTask uploadTask = reference.putFile(fileUri);
        uploadTask.addOnCompleteListener(task -> {
            if (task.isSuccessful()) {
                getDownloadURL(reference, callback);

            } else {
                if (callback != null) {
                    String erEn;
                    String erAr;
                    Exception exception = task.getException();
                    if (exception != null) {
                        erEn = exception.getMessage();
                        erAr = erEn;
                    } else {
                        erEn = "Upload failed";
                        erAr = "تعذر ارسال الملف";
                    }
                    callback.invoke(Response.failed(new StringSet(erEn, erAr)));
                }
            }
        });
    }

    //----------------------------------------------------------------------------

    @Override
    public void getDownloadURL(String firebasePath, ResultCallback<Response<String>> callback) {
        //firebasePath = _refine Path(firebasePath);
        var reference = storage.getReference(firebasePath);
        getDownloadURL(reference, callback);
    }

    public void getDownloadURL(StorageReference reference, ResultCallback<Response<String>> callback) {
        reference.getDownloadUrl().addOnCompleteListener(task1 -> {
            if (task1.isSuccessful()) {
                if (callback != null) {
                    callback.invoke(Response.success(task1.getResult().toString()));
                }
            } else {
                if (callback != null) {
                    String erEn;
                    String erAr;
                    Exception exception = task1.getException();
                    if (exception != null) {
                        erEn = exception.getMessage();
                        erAr = erEn;
                    } else {
                        erEn = "Creating download link failed";
                        erAr = "تعذر انشاء رابط للتحميل";
                    }
                    callback.invoke(Response.failed(new StringSet(erEn, erAr)));
                }
            }
        });
    }

    //----------------------------------------------------------------------------

    @Override
    public void download(String firebasePath, File toDir, @Nullable File suggestedOutFile, ResultCallback<Response<File>> callback) {
        File outFile;

        if (suggestedOutFile != null) {
            outFile = suggestedOutFile;
        } else {
            var i = firebasePath.lastIndexOf('/');
            String fileName;
            if (i >= 0) {
                fileName = firebasePath.substring(i + 1);
            } else {
                fileName = DateOp.getInstance().formatDate(DateOp.PATTERN_yyyy_MM_dd_HH_mm_ss, true).replace(
                        "-",
                        ""
                );
            }

            outFile = new File(toDir, fileName);
        }

        if (!outFile.exists()) {
            try {
                outFile.createNewFile();
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }

        //fromPath = _refine Path(fromPath);
        storage.getReference(firebasePath).getFile(outFile).addOnCompleteListener(task -> {
            if (callback != null) {
                if (task.isSuccessful()) {
                    callback.invoke(Response.success(outFile));
                } else {
                    String erEn;
                    String erAr;
                    Exception exception = task.getException();
                    if (exception != null) {
                        erEn = exception.getMessage();
                        erAr = erEn;
                    } else {
                        erEn = "Creating download link failed";
                        erAr = "تعذر انشاء رابط للتحميل";
                    }
                    callback.invoke(Response.failed(new StringSet(erEn, erAr)));
                }
            }
        });
    }

    //----------------------------------------------------------------------------

    @Override
    public void delete(String atPath, ResultCallback<Response<Boolean>> callback) {
        //atPath = _refine Path(atPath);
        var reference = storage.getReference(atPath);
        reference.delete().addOnCompleteListener(task -> {
            if (task.isSuccessful()) {
                if (callback != null) {
                    callback.invoke(Response.success(true));
                }
            } else {
                if (callback != null) {
                    String erEn;
                    String erAr;
                    Exception exception = task.getException();
                    if (exception != null) {
                        erEn = exception.getMessage();
                        erAr = erEn;
                    } else {
                        erEn = "Creating download link failed";
                        erAr = "تعذر انشاء رابط للتحميل";
                    }
                    callback.invoke(Response.failed(new StringSet(erEn, erAr)));
                }
            }
        });
    }


}

File: GMUtilsFlutter/gmutils_flutter/lib/zgmutils/data_utils/firebase/firebase_storage.dart
import 'dart:io';

import 'package:firebase_core/firebase_core.dart';
import 'package:firebase_storage/firebase_storage.dart' as firebase_storage;
import 'package:path_provider/path_provider.dart';

import '../../utils/date_op.dart';
import '../../utils/logs.dart';
import '../../utils/collections/string_set.dart';
import 'firebase_utils.dart';
import 'fb_response.dart';

abstract class IFirebaseStorage {
  Future<FBResponse<String>> upload(File file, {required String toPath});

  Future<FBResponse<String>> getDownloadURL(String firebasePath);

  Future<FBResponse<File>> download(String fromPath, {File? suggestedOutFile});

  Future<FBResponse<bool>> delete(String atPath);
}

class FirebaseStorage implements IFirebaseStorage {
  late final firebase_storage.FirebaseStorage storage;

  FirebaseStorage() {
    //Firebase.initializeApp();
    storage = firebase_storage.FirebaseStorage.instance;
  }

  String _refinePath(String path) {
    path = FirebaseUtils.refinePathFragmentNames(path);
    return path;
  }

  //----------------------------------------------------------------------------

  @override
  Future<FBResponse<String>> upload(File file, {required String toPath}) async {
    toPath = _refinePath(toPath);
    var ref = storage.ref(toPath);
    return uploadTo(ref, file);
  }

  Future<FBResponse<String>> uploadTo(
    firebase_storage.Reference reference,
    File file,
  ) async {
    Logs.print(() => 'FirebaseStorage[Call].uploadTo('
        'reference: ${reference.fullPath}, '
        'file: ${file.path})');

    try {
      await reference.putFile(file);
      var link = await reference.getDownloadURL();
      Logs.print(() => 'FirebaseStorage[Response].uploadTo(reference: ${reference.fullPath}) ----> $link');
      return FBResponse.success(data: link);
    } on FirebaseException catch (e) {
      Logs.print(() => 'FirebaseStorage[Response.Exception].uploadTo() ----> $e');
      return FBResponse.failed(error: StringSet(e.code));
    }
  }

  //----------------------------------------------------------------------------

  @override
  Future<FBResponse<String>> getDownloadURL(String firebasePath) async {
    Logs.print(() => 'FirebaseStorage[Call].getDownloadURL(firebasePath: $firebasePath)');

    try {
      firebasePath = _refinePath(firebasePath);
      var reference = storage.ref(firebasePath);
      var url = await reference.getDownloadURL();
      Logs.print(() => 'FirebaseStorage[Response].getDownloadURL(firebasePath: $firebasePath) ---> $url');
      return FBResponse.success(data: url);
    } catch (e) {
      Logs.print(() => 'FirebaseStorage[Response.Exception].getDownloadURL ---> $e');
      if (e is FirebaseException) {
        return FBResponse.failed(error: StringSet(e.code));
      } else {
        return FBResponse.failed(error: StringSet('$e'));
      }
    }
  }

  //----------------------------------------------------------------------------

  @override
  Future<FBResponse<File>> download(
    String fromPath, {
    File? suggestedOutFile,
  }) async {
    Logs.print(() => 'FirebaseStorage[Call].download(fromPath: $fromPath)');

    File outFile;
    if (suggestedOutFile != null) {
      outFile = suggestedOutFile;
    } else {
      var i = fromPath.lastIndexOf('/');
      String fileName;
      if (i >= 0 && i < fromPath.length) {
        fileName = fromPath.substring(i + 1);
      } else {
        fileName = DateOp()
            .formatForDatabase(DateTime.now(), dateOnly: false)
            .replaceAll(
              "-",
              '',
            );
      }

      Directory appDocDir = await getDownloadsDirectory() ??
          await getExternalStorageDirectory() ??
          await getApplicationDocumentsDirectory();
      outFile = File('${appDocDir.path}/$fileName');
    }

    if (!outFile.existsSync()) {
      outFile.createSync(recursive: true);
    }

    try {
      fromPath = _refinePath(fromPath);
      await storage.ref(fromPath).writeToFile(outFile);
      Logs.print(() => 'FirebaseStorage[Response].download(fromPath: $fromPath) ---> file downloaded to: ${outFile.path}');
      return FBResponse.success(data: outFile);
    } on FirebaseException catch (e) {
      Logs.print(() => 'FirebaseStorage[Response.Exception].download() ---> $e');
      return FBResponse.failed(error: StringSet(e.code));
    }
  }

  //----------------------------------------------------------------------------

  @override
  Future<FBResponse<bool>> delete(String atPath) async {
    Logs.print(() => 'FirebaseStorage[Call].delete(atPath: $atPath)');

    try {
      atPath = _refinePath(atPath);
      var reference = storage.ref(atPath);
      await reference.delete();
      Logs.print(() => 'FirebaseStorage[Response].delete(atPath: $atPath) -----> true');
      return FBResponse.success(data: true);
    } catch (e) {
      Logs.print(() => 'FirebaseStorage[Response].delete() -----> EXCEPTION:: $e');
      if (e is FirebaseException) {
        return FBResponse.failed(error: StringSet(e.code, e.code));
      } else {
        return FBResponse.failed(error: StringSet('$e', '$e'));
      }
    }
  }
}
